### ISSUE TITLE
Suse: Relocation Error on LEAP

### ISSUE THREAD

**Issue by noelmcloughlin** [saltstack-formulas/samba-formula#26]
Vanilla installation of OpenSuse Leap (server) includes following Samba packages.

```
gvfs-backend-samba-1.28.3-2.2.x86_64
libsamba-credentials0-32bit-4.4.2-7.2.x86_64
libsamba-credentials0-4.4.2-7.2.x86_64
libsamba-errors0-32bit-4.4.2-7.2.x86_64
libsamba-errors0-4.4.2-7.2.x86_64
libsamba-hostconfig0-32bit-4.4.2-7.2.x86_64
libsamba-hostconfig0-4.4.2-7.2.x86_64
libsamba-passdb0-32bit-4.4.2-7.2.x86_64
libsamba-passdb0-4.4.2-7.2.x86_64
libsamba-util0-32bit-4.4.2-7.2.x86_64
libsamba-util0-4.4.2-7.2.x86_64
samba-client-32bit-4.4.2-7.2.x86_64
samba-client-4.4.2-7.2.x86_64
samba-libs-32bit-4.4.2-7.2.x86_64
samba-libs-4.4.2-7.2.x86_64
samba-winbind-32bit-4.4.2-7.2.x86_64
samba-winbind-4.4.2-7.2.x86_64
yast2-samba-client-3.1.18-1.1.noarch
yast2-samba-server-3.1.16-1.2.noarch
```
Executing the **samba.init state** results in broken installation-

`: relocation error: /usr/lib64/samba/libauthkrb5-samba4.so: symbol tevent_req_is_unix_error, version TEVENT_UTIL_0.0.1 not defined in file libtevent-util.so.0 with link time reference`

This is known issue in OpenSUSE (tumbleweed) manifesting in LEAP (version 42).
https://bugzilla.opensuse.org/show_bug.cgi?id=975862

**Comment by noelmcloughlin** [comment]
Possible Solutions
1. Adding a state before **samba.pkg.installed** to clean samba baseline. This works!!
```
{% if grains.os_family == 'Suse' and grains.osmajorrelease in (42) %}
# Prevent relocation error breaking samba on OpenSuse 42.
# See: https://bugzilla.opensuse.org/show_bug.cgi?id=975862
samba_erase32bit:
  pkg.removed:
    - pkgs:
      - libsamba-credentials0
      - libsamba-errors0
      - libsamba-hostconfig0
      - libsamba-passdb0
      - libsamba-util0
      - samba-client
      - samba-libs
      - samba-winbind
      - libsmbconf0
      - libsmbldap0
{% endif %}
```

2. Add  cmd.run ('**zypper dup**' or '**zypper dup samba**') state before pkg.installed.  Possibly works!!

**Comment by noelmcloughlin** [comment]
The `zypper dup` solution ("a distribution upgrade") probably fixes Tumbleweed, but is probably not a valid solution for OpenSuse Major releases.

**Comment by noelmcloughlin** [comment]
Running `zypper --non-interactive dup --no-allow-vendor-change` fixed Suse Leap, but I need to test again.  Rather than address one specific issue, a generic approach is more acceptable as illustrated below.

**defaults.yaml**
```
samba:
  preinstall:
    cmd:
    osreleases: ()
```
**osmap.yaml**
```
Suse:
  service: smb
  client: samba-client

  preinstall:
    # See: https://bugzilla.opersuse.org/show_bug.cgi?id=975862
    cmd: zypper --non-interactive dup --no-allow-vendor-change
    osreleases: (42)
```

**init.sls**
```
{% if samba.preinstall.cmd and grains.osmajorrelease in samba.preinstall.osreleases %}
# If osmap (or pillar) defines a command to run before pkg.install
# and os-major-release matches, run the defined command.
samba_preinstall_cmd:
  cmd.run:
    - name: {{ samba.preinstall.cmd }}
    - require_in: pkg.samba
{% endif %}
```

**Comment by noelmcloughlin** [comment]
An enhancement to salt for similar hook might look like this-

```
samba:
  pkg.installed:
     - name: {{ samba.pkg }}
     - preinstall_cmd: {{ samba.preinstall.cmd }}
     - preinstall_cmd_onlyif:  <some logic>
```

**Comment by noelmcloughlin** [comment]
Retested this code on OpenSuse Leap (all states successful) & Ubuntu 16 (no regression).  If there is no negative feedback I'll prepare a PR.

**Comment by noelmcloughlin** [comment]
Fixed by #28
